File: plugins/bluenimble-plugin-scripting.javascript/src/main/resources/platform/core/ApiOutput.js
/**
  Represents a service execution output/result<br/>
  <strong>Do not call constructor directly</strong><br/>
  Bluenimble container creates it internally and makes it available for your services to use <br/><br/>
  @class ApiOutput
  @classdesc 
*/

/**
  @constructor
  @access private
*/
var ApiOutput = function (proxy) {
	
	this.clazz 	= 'ApiOutput';
	
	this.proxy 	= proxy;
	
	this.name 			= proxy.name ();
	this.timestamp 		= proxy.timestamp ();
	this.contentType 	= proxy.contentType ();
	this.extension 		= proxy.extension ();
	this.length			= proxy.length ();
	
	this.meta			= proxy.meta ();
	this.data			= proxy.data ();
	
	this.set			= function (key, value) {
		proxy.set (key, value);
		return this;
	};
	this.unset			= function (key) {
		proxy.unset (key);
		return this;
	};
	this.get			= function (key) {
		return proxy.get (key);
	};

	this.pipe			= function (out, position, count) {
		if (typeof position == 'undefined') {
			position = 0;
		}
		if (typeof count == 'undefined') {
			count = -1;
		}
		proxy.pipe (out.proxy, position, count);
	};

	this.toInput		= function () {
		return new InputStream (proxy.toInput ());
	};

};

ApiOutput.prototype.fromBytes = function (name, bytes, contentType, extension) {
	return new JC_ApiByteArrayOutput (name, bytes, contentType, extension);
};
ApiOutput.prototype.fromJson = function (json) {
	return new JC_JsonApiOutput (JC_ValueConverter.convert (json));
};
ApiOutput.prototype.create = function () {
	return new JC_JsonApiOutput (Json.object ());
};
ApiOutput.prototype.Defaults = {
	Id 			: JC_ApiOutput_Defaults.Id,
	Timestamp 	: JC_ApiOutput_Defaults.Timestamp,
	Cache 		: JC_ApiOutput_Defaults.Cache,
	Disposition : JC_ApiOutput_Defaults.Disposition,
	Charset 	: JC_ApiOutput_Defaults.Charset,
	Items 		: JC_ApiOutput_Defaults.Items,
	Count 		: JC_ApiOutput_Defaults.Count,
	Expires		: JC_ApiOutput_Defaults.Expires,
	Cast		: JC_ApiOutput_Defaults.Cast,
	Exit		: JC_ApiOutput_Defaults.Exit
};
ApiOutput.prototype.Disposition = {
	Inline 		: JC_ApiOutput_Disposition.Inline,
	Attachment 	: JC_ApiOutput_Disposition.Attachment
};

File: sdks/bluenimble-jvm-sdk/src/main/java/com/bluenimble/platform/api/ApiOutput.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.bluenimble.platform.api;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.Serializable;
import java.util.Date;

import com.bluenimble.platform.json.JsonObject;
import com.bluenimble.platform.streams.Chunk;
import com.bluenimble.platform.streams.StreamDecorator;

public interface ApiOutput extends Serializable {
	
	interface Defaults {
		String Id 			= "id";
		String Timestamp 	= "timestamp";
		String Cache 		= "cache";
		String Disposition 	= "disposition";
		String Charset 		= "charset";
		String Items 		= "items";
		String Count 		= "count";
		String Expires		= "expires";
		String Cast			= "cast";
	}

	interface Disposition 		{
		String Inline 		= "inline";
		String Attachment 	= "attachment";
	}

	String 				name 		();
	Date 				timestamp 	();
	String 				contentType ();

	String 				extension 	();
	long 				length 		();
	
	ApiOutput 			set 		(String key, Object value);
	ApiOutput 			unset 		(String key);
	Object 				get 		(String key);

	JsonObject 			meta 		();
	JsonObject 			data 		();

	void 				pipe 		(OutputStream out, long position, long count) 	throws IOException;
	InputStream 		toInput 	() 												throws IOException;
	
	void 				pipe 		(OutputStream out, StreamDecorator decorator, Chunk... chunks) 			
																					throws IOException;

}
